internal/entitlement/balanceworker/ingesthandler.go

```diff
@@ -4,38 +4,73 @@ import (
 	"context"
 
 	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/hashicorp/go-multierror"
 
 	"github.com/openmeterio/openmeter/internal/entitlement"
 	"github.com/openmeterio/openmeter/internal/event/spec"
 	"github.com/openmeterio/openmeter/internal/productcatalog"
 	"github.com/openmeterio/openmeter/internal/sink/flushhandler/ingestnotification"
+	"github.com/openmeterio/openmeter/internal/watermill"
+	"github.com/openmeterio/openmeter/pkg/slicesx"
 )
 
-func (w *Worker) handleIngestEvent(ctx context.Context, event ingestnotification.EventIngested) ([]*message.Message, error) {
-	affectedEntitlements, err := w.GetEntitlementsAffectedByMeterSubject(ctx, event.Namespace.ID, event.MeterSlugs, event.SubjectKey)
-	if err != nil {
-		return nil, err
+func (w *Worker) handleBatchedIngestEvent(ctx context.Context, event ingestnotification.EventBatchedIngest) ([]*message.Message, error) {
+	filters := slicesx.Map(event.Events, func(e ingestnotification.IngestEventData) IngestEventQueryFilter {
+		return IngestEventQueryFilter{
+			Namespace:  e.Namespace.ID,
+			SubjectKey: e.SubjectKey,
+			MeterSlugs: e.MeterSlugs,
+		}
+	})
+	affectedEntitlements, repoErr := w.connectors.repo.ListAffectedEntitlements(ctx, filters)
+	if repoErr != nil {
+		ev, err := spec.NewCloudEvent(
+			spec.EventSpec{
+				Source: spec.ComposeResourcePathRaw(string(ingestnotification.EventBatchedIngest{}.Spec().Subsystem)), // TODO: this should be a different source
+			},
+			event,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		msg, err := w.opts.Marshaler.MarshalEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		return nil, watermill.NewRetryableError([]*message.Message{msg}, repoErr)
 	}
 
+	failedEntitlementInfo := make([]IngestEventDataResponse, 0)
+	var handlingError error
+
 	result := make([]*message.Message, 0, len(affectedEntitlements))
 	for _, entitlement := range affectedEntitlements {
 		messages, err := w.handleEntitlementUpdateEvent(
 			ctx,
-			entitlement,
+			NamespacedID{Namespace: entitlement.Namespace, ID: entitlement.EntitlementID},
 			spec.ComposeResourcePath(entitlement.Namespace, spec.EntityEvent),
 		)
 		if err != nil {
-			return nil, err
+			// TODO: add error information too
+			handlingError = multierror.Append(handlingError, err)
+			failedEntitlementInfo = append(failedEntitlementInfo, entitlement)
+			continue
 		}
 
 		result = append(result, messages...)
 	}
 
-	return result, nil
+	var err error
+	if len(failedEntitlementInfo) > 0 {
+		err = w.newEntitlementProcessingFailedError(failedEntitlementInfo, handlingError)
```

Comment:
After thinking about this PR a bit, instead of an abstract solution, I would rather include the splitting here, as right now we don't need this anywhere else. If we need the same logic somewhere else we can abstract this, but I wouldn't do it just now.